### DISCUSSION TITLE
how can check 300GB file integrity(checksum) downloaded on network path from s3 payerbucket in c#

### DISCUSSION THREAD

**Discussion by krishsingh-git** [aws/aws-sdk-net#3306] [Q&A] [answered]
how can check 300GB file integrity(checksum) downloaded on network path from s3 payerbucket in c#

getmetadata for each part failed due to multiple bad request aws s3 bucket

**Comment by ashishdhingra** [comment]
@krishsingh-git Good afternoon. Could you please elaborate your scenario, including a code sample and error you are getting, where you specify `getmetadata for each part failed due to multiple bad request aws s3 bucket`? AWS .NET SDK does not support multi-part download as of now. Hence, how are you calling `GetObjectMetadata` for each part!

Thanks,
Ashish

**Comment by krishsingh-git** [comment]
Can we connect on call?i can explain better there

**Comment by krishsingh-git** [comment]
Hi Ashis,

I am facing one more issue ,
I am trying to verify hash value of downloaded file from s3 bucket.
But MD5 value is coming different in local,
Can you please look into my code
"sourceUrl":"s3://nprd-cntdel-wh-s3/VUBIQUITY_UK/Source/615685.mxf"

```
using (var outputStream = new MemoryStream())
            {
                using (MD5 md5 = MD5.Create())
                {
                    using (FileStream fileStream = new FileStream(MD5FileIntegrity.Path.Value, FileMode.Open, FileAccess.Read))
                    {
                        byte[] bufferLast = new byte[0];
                        int i = 1;
                        while (i <= MD5FileIntegrity.NumberOfParts.Value)
                        {
                            Result<GetObjectMetadataResponse> partMetadataResult = await _s3Service.GetObjectMetaDataRetryTokenAsync(s3Location, i, amazonS3, s3ExtendedCopyClient);
                            long chunkSize = partMetadataResult.Value.ContentLength;

                            byte[] buffer = new byte[(int)chunkSize];
                            bufferLast = new byte[(int)chunkSize];
                            var bytesRead = await fileStream.ReadAsync(buffer, 0, buffer.Length).ConfigureAwait(false);
                            var bytesReads =  fileStream.Read(bufferLast, 0, bufferLast.Length);
                            Console.WriteLine("bytesRead is " + bytesReads);
                            md5.TransformBlock(buffer, 0, bufferLast.Length, null, 0);
                            outputStream.Write(buffer, 0, bytesReads);
                            i++;
                        }
                        md5.TransformFinalBlock(outputStream, 0, outputStream.Length);
                    }

                    var myChecksum = BitConverter.ToString(md5.Hash).Replace("-", "").ToUpper();
```

**Reply by ashishdhingra** [reply, marked as answer]
@krishsingh-git You could try checking [Checking object integrity](https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html) and [New – Additional Checksum Algorithms for Amazon S3](https://aws.amazon.com/blogs/aws/new-additional-checksum-algorithms-for-amazon-s3/).

SDK(s) also have support for flexible checksums.
- If the object was uploaded with flexible checksums, users should be able to set `getObjectRequest.ChecksumMode = ChecksumMode.Enabled` to trigger the SDK to validate the integrity as the response stream is read.
- BUT, we skip that if the object was multi-part uploaded, since the checksum is actually a checksum-of-checksums (SDK has the check [here](https://github.com/aws/aws-sdk-net/blob/24c0f425f8df008dac0b5fad3d208b83073ac667/sdk/src/Core/Amazon.Runtime/Internal/Util/ChecksumUtils.cs#L194))
- So you would need to validate checksums on a part-by-part basis.

Thanks,
Ashish
